### ISSUE TITLE
Update CONTRIBUTING.md to improve commit messages and release notes

### ISSUE THREAD

**Issue by notmandatory** [bitcoindevkit/bdk#1521]
Update CONTRIBUTING.md to better explain how to create a PR with a concise title suitable for including in the GitHub automatically generated release notes. Discuss below what a suitable title should look like, but should fit on one line.

Remove "Changelog notice" section from `pull_request_template.md`.

**Comment by notmandatory** [comment]
I think we should also discuss what to do if a PR's changes can't easily be explained in one brief line.  I think this is a good indication that a PR is trying to change too much and should be broken up into multiple PRs.

**Comment by ValuedMammal** [comment]
Agree, PRs should be focused with a well defined scope and avoid introducing changes that aren't relevant to the issue it purports to resolve.

If a PR only has one commit then it makes sense to use the same title as the commit message for the PR title (in conventional commits style). If a PR has more than one commit, how should the PR title appear? just name it after the most relevant commit?

In terms of auto generated release notes, should the PR title read more like a commit message or a changelog entry? or is there not much difference? For example I noticed contributors (myself included) sometimes opt for a more high-level description of the change as opposed to making a direct reference to code and implementation. In any case it would help for the title of a PR to include the type of change (doc, fix, feat, refactor, example, chore) as well as the crate or module scope.

**Comment by notmandatory** [comment]
As part of the update to this file I'd also like to propose PR authors and reviewers are responsible for bumping workspace crate versions appropriately when a breaking change is needed. For example, if a breaking change is made to `bdk_chain` then: 

1. bump the crate version of `bdk_chain` to next 0.x.0 minor version.
2. bump the `bdk_chain` version in the dependency section of `bdk_wallet`, `bdk_file_store`, `bdk_testenv`.
3. bump the crate versions to next 0.y.0 versions for `bdk_wallet`, `bdk_file_store`, `bdk_testenv`.
4. add or update the workspace cargo.toml `[patch crates-io]` so the workspace will build with unreleased versions:
   ```toml
   [patch.crates-io]
   bdk_chain = { path = 'crates/chain' }
   ```

When a release is published the person doing the release doesn't need to remove the `[patch.crates-io]` section. No files need to be edited only the release tag added.  We might even be able to remove the `path = "..."` sections of inter-workspace dependencies which would force devs to manually update versions when making changes.

**Comment by oleonardolima** [comment]
> I propose that we depend on commit descriptions to mark changes as breaking/non-breaking. PR authors and reviewers are responsible for enforcing commit descriptions contain a `!` when necessary. Our commit history is already very clean in this regard - we follow the [Conventional Commit Specification](https://www.conventionalcommits.org/en/v1.0.0/) very well.
> 
> To figure out if a certain crate has breaking changes, we can iterate through the commit descriptions from the last release, find all commits with !, and the associated crates that are changed. A tool that can automatically do this will be great.

I guess if those commits/PR titles have the affected crate as the scope in `<type>[optional scope]: <description>` semantic commit style, it can make it even easier to spot and know which crate needs which bump.

**Comment by oleonardolima** [comment]
I'd like to share that I found this tool https://git-cliff.org/docs/usage/bump-version (it's been mentioned in today's TWiR newsletter). It looks like it is fairly customizable and also can handle the conventional commits pretty well, I guess it can be really helpful.
